### ISSUE TITLE
Index to Callchain-array 0 instead of 1 in some of the Action Vis

### ISSUE THREAD

**Issue by fuzzy-focus** [fieldrndservices/logger#3]
In some of the reporting action VIs (trace, debug, info, ...) and the catch error VI, the source of the message is set to the 0th element of the callchain, which is the VI from the logging library.
![image](https://user-images.githubusercontent.com/13691063/62440835-17711900-b752-11e9-817e-8472faaab7f1.png)

I assume this should have been the VI that is issuing the report, which would be the first element.

**Comment by volks73** [comment]
Normally, yes the zeroth element in the call chain would be the VI where the Call Chain function is used, i.e. the Fatal, Error, Warning, Info, Debug, and Trace VIs. However, each of these VIs are "inlined" into the block diagram of the VI calling the action VI to reduce sub-VI execution overhead and improve performance. Inlining a VI is essentially copying-and-pasting the contents of the sub-VI into the parent, calling VI's block diagram as if the sub-VI did not exist. The [VI Execution Speed](https://zone.ni.com/reference/en-XX/help/371361R-01/lvconcepts/vi_execution_speed/) reference document from [NI](http://www.ni.com) explains the inlining process under the "SubVI Overhead" section. From this document:

>A third way to minimize subVI overhead is to inline subVIs into their calling VIs. When you inline a subVI, LabVIEW inserts the compiled code of the subVI into the compiled code of the calling VI. If you then make changes to the subVI, LabVIEW recompiles all calling VIs of that subVI to include those changes. Essentially, inlining a subVI removes the need to call the subVI at run time. Instead, LabVIEW executes the subVI code inside the compiled code of the calling VI.
>
>Inlining subVIs is most useful for small subVIs, subVIs within a loop, subVIs with unwired outputs, or subVIs you call only once. To inline a subVI, place a checkmark in the **Inline subVI into calling VIs** checkbox on the [Execution](https://zone.ni.com/reference/en-XX/help/371361R-01/lvdialog/execution/) page of the VI Properties dialog box. You must also select **Preallocated clone reentrant execution** on the same page of the dialog box. LabVIEW automatically preallocates clones for each instance when it inlines the subVI.

Thus, the action and error handling VIs within the compiled code is actually in the parent, calling VI's compiled code and the Call Chain function will return the parent, calling VI instead of the action/error VIs for the zeroth element.

This through me for a loop when I first started using the Call Chain function and trying to inline sub-VIs.

**Comment by fuzzy-focus** [comment]
All you are saying does make a lot of sense, inlining these VIs is very sensible and then the callchain issue is resolved.

But the reason why I created this issue was the fact that the VIs using the callchain were appearing in my logfiles, and those entries are quite wide.
So I'm guessing the inlining does not work properly?

**Comment by volks73** [comment]
Hmm, the long call chains are probably caused by deeply nested VIs. I run into this quite frequently since I tend to use a lot of sub-VIs and project libraries. 

Would it be possible to post an image or an example log file with the excessively wide Source fields?

Changing the usage of the Call Chain function within the Action and Error VIs is probably not going to be the appropriate solution for excessively wide Source fields in the log files. The Call Chain acts like a stack trace to see where exactly the log statement originated. In deeply nested VIs, you probably do not need the full "stack trace" to locate the log statement within the codebase. 

So, there might need to be some configuration, either global to the Logger package or local to each Action and Error VI, that truncates the Source field to some number of nested VIs. For example, a nesting of 5 would show up to the 5 parent VIs from the Action or Error VIs. However, the Source field could still be excessively wide because of the fully qualified VI names, which includes project libraries. An additional option to use the common name or short name would be needed, but now we are starting to get a lot of configurations and options. There are probably more things that need to be implemented and toggled for this to work in every situation. We start losing the simplicity of the package.

Another option is to change the Source to a string control. If the string is empty, then the Call Chain function is used to populate the Source, but if a string is wired to the control, then it is used directly and the Call Chain function is skipped. The Source field would have to be surrounded in double quotes and escaped like the message field. This might be the easier and more flexible option. Then, for your deeply nested VIs, you can just pass the parent's name or any codebase meaningful source value to the Action and Error VIs.

I will have to think about this, and some other idea may present itself if you provide an example of the excessively wide Source fields.

**Comment by fuzzy-focus** [comment]
I think this was a miscomunication on my side. with excessively long fields I didn't mean in prints the whole stacktrace. I "just" get "LibraryName:VI-name"  of the Action and Error VIs (I don't know the full library name right now, but its basically something like "FieldRND_Logger.lvlib:Debug.vi", it's not extremely long but it does take up considerable screen real estate).

This would suggest that the inlineing wasn't working right.

I will give more information tomorrow when I have my work files available.

**Comment by volks73** [comment]
Interesting. Can you run the "Simple.vi" example and note the output for the messages?

When I run this example on my computer (LabVIEW 2018), only get "Simple.vi" for the source field. For example,
```text
...
20190805T143701.186, TRACE, Simple.vi, "Iteration begins"
20190805T143701.186, DEBUG, Simple.vi, "Iteration: 40"
...
```
Can you verify that the Debug VI in your version is inlined?

**Comment by fuzzy-focus** [comment]
When I run the same VI I get this output:

```
Timestamp, Tag, Source, Message
20190806T152126,555, TRACE, Field_RnD_Services_Logger_Toolkit.lvlib:Trace.vi:6130001, "Iteration begins"
20190806T152126,555, DEBUG, Field_RnD_Services_Logger_Toolkit.lvlib:Debug.vi:4190001, "Iteration: 0"
20190806T152126,555, INFO, Field_RnD_Services_Logger_Toolkit.lvlib:Info.vi:2430001, "Even numbered iteration occurred"
20190806T152126,555, TRACE, Field_RnD_Services_Logger_Toolkit.lvlib:Trace.vi:6130002, "Iteration ends"
20190806T152126,805, TRACE, Field_RnD_Services_Logger_Toolkit.lvlib:Trace.vi:6130001, "Iteration begins"
20190806T152126,805, DEBUG, Field_RnD_Services_Logger_Toolkit.lvlib:Debug.vi:4190001, "Iteration: 1"
20190806T152126,805, WARN, Field_RnD_Services_Logger_Toolkit.lvlib:Warn.vi:2140001, "Odd numbered iteration occurred"
20190806T152126,805, TRACE, Field_RnD_Services_Logger_Toolkit.lvlib:Trace.vi:6130002, "Iteration ends"
20190806T152127,055, TRACE, Field_RnD_Services_Logger_Toolkit.lvlib:Trace.vi:6130001, "Iteration begins"
20190806T152127,055, DEBUG, Field_RnD_Services_Logger_Toolkit.lvlib:Debug.vi:4190001, "Iteration: 2"
20190806T152127,055, INFO, Field_RnD_Services_Logger_Toolkit.lvlib:Info.vi:2430001, "Even numbered iteration occurred"
20190806T152127,055, TRACE, Field_RnD_Services_Logger_Toolkit.lvlib:Trace.vi:6130002, "Iteration ends"
20190806T152127,305, TRACE, Field_RnD_Services_Logger_Toolkit.lvlib:Trace.vi:6130001, "Iteration begins"
20190806T152127,305, DEBUG, Field_RnD_Services_Logger_Toolkit.lvlib:Debug.vi:4190001, "Iteration: 3"
20190806T152127,305, WARN, Field_RnD_Services_Logger_Toolkit.lvlib:Warn.vi:2140001, "Odd numbered iteration occurred"
20190806T152127,305, TRACE, Field_RnD_Services_Logger_Toolkit.lvlib:Trace.vi:6130002, "Iteration ends"
```

I have updated to the latest version (Logger v1.9.0.73).

When I open one of the action VIs and check the execution settings I get this:
![image](https://user-images.githubusercontent.com/13691063/62543569-6ad08d80-b85e-11e9-9015-eb3f8949d2d6.png)


So the inlining is not set.

After manually setting the inline checkbox on all the relevant VIs it works:

```
Timestamp, Tag, Source, Message
20190806T152728,111, TRACE, Simple.vi, "Iteration begins"
20190806T152728,111, DEBUG, Simple.vi, "Iteration: 0"
20190806T152728,111, INFO, Simple.vi, "Even numbered iteration occurred"
20190806T152728,111, TRACE, Simple.vi, "Iteration ends"
20190806T152728,361, TRACE, Simple.vi, "Iteration begins"
20190806T152728,361, DEBUG, Simple.vi, "Iteration: 1"
20190806T152728,361, WARN, Simple.vi, "Odd numbered iteration occurred"
20190806T152728,361, TRACE, Simple.vi, "Iteration ends"
20190806T152728,611, TRACE, Simple.vi, "Iteration begins"
20190806T152728,611, DEBUG, Simple.vi, "Iteration: 2"
20190806T152728,611, INFO, Simple.vi, "Even numbered iteration occurred"
20190806T152728,611, TRACE, Simple.vi, "Iteration ends"
```

apparently something in the packaging process unchecks these boxes.

**Comment by volks73** [comment]
Thanks for the information. I figured out the source of the problem and created a new issue, #6, to track its progress. Thanks again!
